### ISSUE TITLE
Consider creating coordinate frames that map on to standard velocity frames

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#10212]
As part of the work on https://github.com/astropy/astropy/pull/10185, the issue came up a few times about how we should deal with different velocity frames in astropy, which aren't already covered by existing coordinate frames. For example, while ICRS can be used to represent a frame that is stationary relative to the solar system barycenter, we don't have any frames at the moment which co-move with the LSRK or LSRD velocity frames or other commonly used frames (e.g. the WCS standard also defines a galactocentric velocity frame, a local group velocity frame, and a CMB dipole velocity frame).

So in https://github.com/astropy/astropy/pull/10185, we create specific coordinate instances that can represent such velocity frames:

```python
    LSRK_GORDON1975 = FK4(x=0 * u.km, y=0 * u.km, z=0 * u.km,
                          v_x=-20 * u.km / u.s * np.cos(270 * u.deg) * np.cos(30 * u.deg),
                          v_y=-20 * u.km / u.s * np.sin(270 * u.deg) * np.cos(30 * u.deg),
                          v_z=-20 * u.km / u.s * np.sin(30 * u.deg),
                          representation_type='cartesian',
                          differential_type='cartesian',
                          equinox='B1900')
```

In fact, this is a specific point in space, but we can use it as a proxy for defining the LSRK reference frame. The downside of this is that the user will never actually see the words 'LSRK' spelt out when introspecting ``SpectralCoord`` objects:

```python
    >>> sc_ttau.with_observer_in_velocity_frame_of(sc_ttau.LSRK_GORDON1975)  # doctest: +REMOTE_DATA +FLOAT_CMP
    <SpectralCoord [200.01903338, 210.01998505, 220.02093672, 230.02188838,
                    240.02284005, 250.02379172, 260.02474339, 270.02569506,
                    280.02664673, 290.0275984 , 300.02855007] GHz
        observer:
          <FK4 Coordinate (equinox=B1900.000, obstime=2019-04-24T02:32:10.000): (x, y, z) in m
              (-1.27820157e+11, -7.20621074e+10, -3.12431755e+10)
           (v_x, v_y, v_z) in km / s
              (1.51824951e-05, 17.32048573, -9.99998236)>
        target:
          <ICRS Coordinate: (ra, dec, distance) in (deg, deg, pc)
              (65.497625, 19.53511111, 144.321)
           (pm_ra_cosdec, pm_dec, radial_velocity) in (mas / yr, mas / yr, km / s)
              (1.37949782e-15, 1.46375638e-15, 23.9)>
        observer to target (computed from above):
          radial_velocity=12.506991149252126 km / s
          redshift=4.171970213029397e-05>
```

Instead, they see an FK4 object. Just to put a couple of possible solutions out there for discussion:

**Option 0:** Status quo, don't change anything, which means no additional development but will almost certainly frustrate users.

**Option 1:** In principle we could create a whole new set of classes that represent pure velocity frames with no origin, and to somehow allow velocities to optionally be defined in a separate frame than the location, so for example:

```python
          <ICRS Coordinate: (ra, dec, distance) in (deg, deg, pc)
              (65.497625, 19.53511111, 144.321)
           (v_x, v_y, v_z) in km/s
              LSRK: (1.0, 2.2, -3.1)>
```

The default would be the current behavior of interpreting them in the same frame of reference. This has the benefit that a user can easily say e.g. 'I want to set up an object with Galactocentric coordinates XYZ and with a velocity relative to LSRD of VXVYVZ', a task that is extremely difficult for anyone but the astropy.coordinates developers at the moment.

**Option 2:** An alternative is that we could consider using regular coordinate frames (perhaps based on ICRS) that differ only in their velocities to defined velocity frames, so for example we could define ``ICRS_LSRK`` or ``ICRS_LSRK_GORDON1975`` which is the ICRS coordinate system which moves relative to ``ICRS`` using the ``LSRK`` velocities. This is not a perfect solution because really the velocities aren't ICRS any more than Galactic - they are just defined as an absolute velocity frame. But it is an option.

I though I'd open this so that we can discuss these options and other options! If you suggest another option, please number it so we can keep track :)

**Comment by keflavich** [comment]
Personally I like either Option 1 or 2.  This is a transformation I need to perform frequently, so seeing it spelled out in code in clear language is helpful.  Is `with_observer_in_velocity_frame_of` still going to be the right way to transform between two velocity frames?  I don't really understand the language as is... the observer is always in the observer's frame, shouldn't the target object be given a velocity frame attribute of, e.g., LSRK, then transform to the observer frame of, say, ITRF?

**Comment by astrofrog** [comment]
The name ``with_observer_in_velocity_frame_of`` is just what's in #10185, feel free to comment there if you have better suggestions for changing the velocity frame of the observer. Interestingly when I think of velocity transforms I think of changing the observer's **velocity frame** (not positional frame) to be LSRK, not changing the target to be in ITRF. Of course the two are equivalent, but I'm not sure what the cleanest API to present this is. We could add for instance ``with_target_in_velocity_frame_of`` and then you could call ``sc.with_target_in_velocity_frame_of(sc.observer)`` to do what you are mentioning.

**Comment by keflavich** [comment]
Right, the example you're invoking makes more sense, but the syntax still feels weird to me.  "with_observer_in...." sounds in English like you're moving the observer, but you're actually changing frames.  It would read more naturally as, "observed_in_velocity_frame_of", but I'm not sure if there are other reasons to avoid that language.

**Comment by adrn** [comment]
@astrofrog Are these conceptually different from the [LSR](http://docs.astropy.org/en/latest/api/astropy.coordinates.builtin_frames.LSR.html#astropy.coordinates.builtin_frames.LSR) frame?

**Comment by astrofrog** [comment]
@adrn - whoa how did I miss that there was an LSR frame... That is exactly what I mean, but we should add the different LSR frames that exist and are used e.g. by the WCS standard as well as e.g. CASA and SLALIB, which includes for example LSRK and LSRD as I defined in #10185. But at least it's good there is precedent in terms of the current LSR frame. So that's basically Option 2 that I listed above. Are we happy to add more similar frames? (If so, I'll add the ones needed for #10185 in that PR)
